### PR TITLE
perf(glm52): graph the dspark draft round — markov chain + piecewise forward (−6.8% bs=1)

### PR DESCRIPTION
**Pull request by n-WN**
## What

Three launch-overhead reductions on the DSpark draft round (#582 items 1–2), measured on an H100 with the single-GPU synthetic smoke; all draft-token-exact:

| H100, bs=1 | round | notes |
|---|---|---|
| all-eager baseline | 2.633 ms | 69% of the 1.82 ms weight-read floor |
| + graphed Markov chain & widened argmax | 2.606 ms | 7-step chain replays as one graph launch (~21 launches collapsed); step-argmax partial 9.7 → 4.5 µs (nsys-verified) |
| + piecewise forward graph | **2.455 ms (−6.8%)** | **74% of floor**; recovers ~the ~158 µs of launch gaps the kernel trace showed |

bs=8: the Markov-chain graph applies (−0.7%); the forward graph falls back to eager (see Limitations).

## How

- **Graphed Markov chain** (`markov_propose`): everything in the 7-step loop is pointer- and shape-static for a given row count — `step` bakes per node, the prev/next ping-pong alternates deterministically, the anchor h2d and sampled d2h stay outside. One `CudaGraphState` per row count; the first round per count runs the plain loop so cuBLAS's lazy workspace allocation stays outside capture. `DSPARK_NO_MARKOV_GRAPH=1` kill-switch.
- **`MARKOV_STEP_TILE_ELEMS 4096 → 1024`** (markov-only): the step argmax reads just one logits row + one bias row (~0.6 MB), so at 4096-elem tiles its ~38 blocks are pure latency (~10 µs for a ~0.2 µs read). 152 blocks put it back on the bandwidth curve. The global-vocab-index tie-break is unchanged, so results are exact; the batched full-row argmax keeps its wider tile.
- **Piecewise forward graph** (`propose`): the only round-varying kernel arguments are `committed_len`-derived — rope positions, the two KV-append copy offsets, attention kv_len — 4 launches per layer, which stay eager (a captured FlashInfer prefill bakes its KV iteration count). The rest is shape-static per `context_len`: captured as `DSPARK_LAYERS + 1` dense segments (embed+fc+L0-head; per-layer tail+next-head; last tail + final norm + lm_head), keyed by `context_len`, lazily captured with a one-round eager warm per key. Host metadata (`set_context_len`, `pending.seq_len`, the tail seq_lens) is hoisted/kept host-side so replays never depend on it. `DSPARK_NO_FORWARD_GRAPH=1` kill-switch.

## Tests

- `dspark_propose_smoke` (cherry-picked from #585 per the #582 discussion, adapted to the `cache_len` plumbing) — the perf A/B fixture.
- **`dspark_propose_graph_parity`** (new, `#[ignore]` GPU): pseudo-random non-degenerate weights, 12 rounds, asserts the graphed path emits exactly the eager path's draft tokens round-for-round, and that the logits are non-degenerate (not the all-ties case).

## Limitations / follow-ups

- The forward graph is **bs=1 only**: the varlen tail scratch (`tail_input`/`k_tail`/`v_tail`) is shared across slots, so the per-slot prep/consume interleave can't be re-ordered into dense-vs-dynamic spans without per-slot tail buffers. bs>1 falls back to eager (Markov graph still applies).
- The `DSPARK_FUSED_MARKOV` cooperative-kernel experiment from the #582 thread is left out (measured negative result; record in the issue).
- Remaining gap to floor is dominated by the 8-row layer GEMMs (~0.25 ms over their weight floor) — cuBLAS shape / TGV territory (#582 item 4).

Closes nothing; part of #582.

### CODE REVIEW

File: openinfer-glm52/src/dspark.rs [hunk 1]
```diff
@@ -399,194 +487,250 @@ impl Glm52DsparkModel {
...
+                            state.committed_len + tail_len,
+                        )?;
+                    }
+                }};
+            }
+            macro_rules! run_seg {
+                ($idx:expr, $body:block) => {{
+                    if use_graph {
+                        forward_graphs[fwd_key][$idx]
+                            .run_or_capture(ctx, || -> Result<()> { $body Ok(()) })?;
```
[the remark this answers is not in the document]
Comment (n-WN) [reply openinfer-glm52/src/dspark.rs]: Correct, and verified — the captured segments bake the slot's `pending`/`context_projected`/`context_hidden` pointers, so a slot rotation at `active == 1` would replay slot A's context projection into slot B's round. Fixed by keying the graphs on **(device address of the slot's `pending` buffer, context_len)** — stable for the slot's lifetime, unlike the `&mut` reference — bounded by slots × block entries. The `dspark_propose_graph_parity` gate now rotates two slots on alternating rounds to pin exactly this case: fails on the old key, passes now (12.8 s, token-exact vs eager). Perf unchanged (bs=1 2.46 ms).

**Comment by xiaguan** [comment]
@codex review

File: openinfer-glm52/src/dspark.rs [hunk 2]
```diff
@@ -399,194 +487,262 @@ impl Glm52DsparkModel {
...
-                )?;
+            // The dense head of layer `l`: input norm, batched q GEMM, and the
+            // per-state tail assembly + k/v GEMMs (constant offsets/shapes for
+            // a given context_len).
+            macro_rules! layer_head {
+                ($l:expr) => {{
+                    let layer = &self.layers[$l];
+                    rms_norm_batch_into(ctx, hidden, &layer.input_ln, DSPARK_RMS_EPS, normed);
+                    gemm_rows_into_checked(ctx, &layer.qkv, 0, DSPARK_QKV_DIM, normed, q_batch)?;
+                    for (i, state) in states.iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (n-WN) [reply openinfer-glm52/src/dspark.rs]: Real — and worse than stated: the restructure had put *every* path (not just the graphed one) into prep-all-then-consume-all order, so plain eager bs>1 was corrupted too (the zero-weight smoke masked it). Fixed in b4cdeda by composing the layer from `layer_prolog` / `state_prep` / `state_dynamic` / `layer_tail`: bs=1 keeps the graphable dense-vs-dynamic split, bs>1 restores the original per-slot interleave. Per-slot tail buffers were considered and rejected — they're sized to `cache_len` (~118 MB per slot). New gate `dspark_propose_batched_slot_isolation`: slot A's drafts must be **bit-identical** across two bs=2 runs differing only in slot B's content (exact by construction, same batch shape — unlike bs-vs-bs1 equality, which cuBLAS's per-M algorithm choice doesn't promise; I tried that first and it legitimately diverges in the last ulp). Passes now.

File: openinfer-glm52/src/dspark.rs [hunk 3]
```diff
@@ -609,6 +765,59 @@ impl Glm52DsparkModel {
...
+                prev_tokens,
+                next_tokens,
+                sampled_tokens,
+                partial_values,
+                partial_indices,
+                ..
+            } = scratch;
+            markov_graphs[rows - 1].run_or_capture(ctx, || {
+                for step in 1..block {
+                    let (prev, next): (&CudaSlice<u32>, &mut CudaSlice<u32>) = if step % 2 == 1 {
```
[the remark this answers is not in the document]
Comment (n-WN) [reply openinfer-glm52/src/dspark.rs:762]: Also real — the eager fallback's 7 field swaps per round (odd) flip which buffer `prev_tokens` names, so an eager warm round for one row count desynced the anchor h2d from a graph captured for another row count. Fixed in b4cdeda: the fallback now uses the same fixed-orientation parity ping-pong as the captured closure and the fields are never swapped, so every capture and every h2d see the same orientation regardless of row-count interleaving. Smoke/parity/isolation all green (bs=1 2.46 ms unchanged).

**Comment by xiaguan** [comment]
@codex review

**Review by xiaguan — CHANGES_REQUESTED** [review]
Thanks — the direction is right (#582's launch overhead is real, and the piecewise split around the committed_len-derived launches is the correct shape), and the measured win reproduces on our reference box. But there are three correctness bugs, one of which kills the engine within seconds of real serving traffic; verified live on jz-38 (8×H200).

## Re-measured perf (jz-38 H200, your smoke, 3 runs each)

| p50/round | graphs on | graphs off | Δ |
|---|---|---|---|
| bs=1 | 1.877 / 1.878 / 1.878 ms | 1.972 / 1.989 / 1.972 ms | **−4.9%** |
| bs=8 | ~4.73 ms | ~4.73 ms | flat |

`dspark_propose_graph_parity` passes on H200. (H200's floor is lower than your H100 numbers across the board, so the relative win shrinks from −6.8% to −4.9%.)

## Blocking bugs

**1. The forward-graph replay reads stale `k_tail/v_tail/tail_input.seq_len` — engine teardown within a few requests.** The `seq_len` assignments sit inside `layer_head!`, i.e. inside the captured segment (`dspark.rs:567-569`); replays don't run the closure, so the always-eager `layer_dynamic!` consumes whatever the last capture/eager round left there. `context_len` is the previous round's accept length (1..=span), so mixed accept lengths poison it immediately. **Live crash on jz-38** during a plain c1 sharegpt bench, second request in: `thread 'glm52-rank-0' panicked at openinfer-kernels/src/ops/elementwise.rs:294: source token range [0..12) exceeds seq_len 11` → draft response dropped → whole-engine teardown. (The stale-too-large direction doesn't assert but ropes garbage rows, with an OOB rope-table read once positions near the cache cap.) Fix: hoist these three `seq_len` writes out of the captured region, next to the `state.pending.seq_len` hoist you already did.

**2. The eager Markov loop's `mem::swap(prev_tokens, next_tokens)` (7 swaps = net flip) silently poisons every already-captured Markov graph.** The graph path selects by `step % 2` and never swaps; the anchor h2d writes whatever `prev_tokens` currently points at. First warm (eager) round of a NEW row count flips the fields, and every previously captured count's replay then embeds from the wrong buffer — drafts go wrong silently, verify corrects them, and the symptom is accept rate collapsing to the floor under multi-slot load. Fix is also the cleanup: delete the eager path's swap and use the same parity-selection body as the capture closure.

**3. `MARKOV_STEP_TILE_ELEMS` 4096→1024 breaks qwen3's DFlash Markov head at runtime.** `markov_step_argmax_into` now checks `markov_step_argmax_partials_len` (4× entries), but `openinfer-qwen3/src/dspark.rs:180,216` still sizes `MarkovScratch` with `argmax_batch_bf16_split_partials_len` — every qwen3 markov call now fails "markov step partials too small" (and that check is the only thing standing between this and a device OOB). Both qwen3 sizing sites (incl. `reservation_bytes`) need the new function.

## Also needed

- The parity test can't see bugs 1 and 2 by construction: it appends exactly 1 row per round (constant `context_len`) and never varies the row count. Add mixed accept lengths (append 1..8 rows) and a mixed-rows interleave (a 2-active round between bs=1 rounds).
- `dspark.rs` is now 1138 lines — over this repo's 1k ceiling; `synthetic()`/`randomize_for_test()` moving into `dspark_smoke.rs` gets it back under.
- The smoke's VRAM comment says ~11 GB / "slot states 2.8 GB"; the arena math says ~5.5 GiB for bs=8 states and ~13.5 GiB peak — please re-derive.
- Nits: `set_tail_len` is now dead; `if fwd_on && use_graph` is just `if use_graph`; the smoke doc's build command (`-p openinfer-glm52 --features glm52`) doesn't exist on this package — drop the feature flag; the parity test's `unsafe set_var` assumes `--test-threads=1` with no enforcement.

Happy to re-run the jz-38 A/B once these land.

**Comment by n-WN** [comment]
All three addressed — thank you for the live verification, that stale-`seq_len` crash is exactly the class of bug my constant-context smoke could never see.

**1. Stale tail `seq_len` on replay** — real, and my miss: I hoisted `pending.seq_len` out of the captured region but left the three tail bounds inside. They're now set in the host-prep section before the segments (bs=1's single assignment is the round's truth; at bs>1 `state_prep` still re-sets them per slot, eagerly). 

**2. Eager Markov swap** — fixed in b4cdeda a few minutes before your review landed (same prescription: the fallback now uses the capture closure's parity selection and the fields are never swapped). Codex's re-review on that commit is what the "chef's kiss" was about.

**3. qwen3 DFlash Markov sizing** — real, and the worse miss on my side: I changed a shared kernel's contract without grepping all callers. Both qwen3 sizing sites (scratch alloc and `MarkovScratch::bytes` behind `reservation_bytes`) now use `markov_step_argmax_partials_len`; `openinfer-qwen3` compiles against this branch and its lib tests pass 64/64.

**Also-needed items, all in:**
- The parity gate now runs a serving-shaped schedule: mixed accept lengths (each `context_len` its own graph key), two bs=2 rounds interleaved between bs=1 rounds, and — the tooth that matters for bug 1 — a replay of the `context_len == 8` graph scheduled right after short-tail rounds. I verified the gate has teeth empirically: with only the `seq_len` hoist reverted, it dies at that replay round with `source token range [0..16) exceeds seq_len 11` at the same `elementwise.rs` range check as your engine-teardown crash; with the hoist restored the full schedule passes. (This cycle also cost me an evening to a subtler trap: `rsync -a` preserves mtimes, so restoring the fixed file gave it an *older* mtime than the reverted one and cargo silently kept the stale binary — `touch` + `cargo clean -p` before the positive run.)
- `dspark.rs` is back under the ceiling (983 lines): `synthetic`/`randomize_for_test` and the `accept_prefix_match` tests moved to a `#[cfg(test)]` child module, and `Glm52DsparkSlotState` moved to a `slot` child module (`pub(super)` fields).
- VRAM comment re-derived per the arena math (~14 GiB peak); the stale `--features glm52` build command dropped; dead `set_tail_len` removed; `if fwd_on && use_graph` simplified; the env-mutating tests now serialize on a local `ENV_LOCK` mutex instead of assuming `--test-threads=1`.

All of the above is in 8656110. H100 gates on that head: upgraded mixed-schedule parity, batched slot isolation, smoke (bs=1 2.457 ms avg — unchanged from the pre-review head) — all green. Would appreciate the H200 A/B re-run when you have a moment.

**Comment by xiaguan** [comment]
@codex reviewe

**Comment by xiaguan** [comment]
@codex review
